staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -338,6 +355,29 @@ func New(opts Options) (*Authenticator, error) {
 	return authenticator, nil
 }
 
+// discoveryURLRoundTripper is a http.RoundTripper that rewrites the issuer URL to the
+// discovery URL for the /.well-known/openid-configuration endpoint.
+type discoveryURLRoundTripper struct {
+	base         http.RoundTripper
+	discoveryURL *url.URL
+}
+
+func (t *discoveryURLRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If the discovery URL is set in authentication configuration, we should
+	// use the discovery URL to fetch the openid configuration. This is useful
+	// for self-hosted providers, for example, providers that run on top of
+	// Kubernetes itself.
+	if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, wellKnownEndpointPath) {
+		if t.discoveryURL != nil {
+			clone := net.CloneRequest(req)
+			clone.URL.Host = t.discoveryURL.Host
+			clone.URL.Path = strings.TrimSuffix(t.discoveryURL.Path, "/") + wellKnownEndpointPath
```

Comment:
Once you make it so discovery URL cannot end in a slash, this can be simplified:

```suggestion
			clone.URL.Path = t.discoveryURL.Path + wellKnownEndpointPath
```